<div class="alert alert-danger">
    <h3>NOTE</h3>
    <p>Before you submit this assignment, <strong>make sure everything runs as expected</strong>:</p>
    <ol>
        <li><strong>restart the kernel</strong> (in the menubar, select <strong>Kernel → Restart</strong>)
        <li><strong>run all cells</strong> (in the menubar, select <strong>Cell → Run All</strong>)</li>
    </ol>
    <p>Make sure to complete every cell that states "<strong><TT>YOUR CODE IN THIS CELL</TT></strong>".</p>
</div>

# Assignment #2

<font color=red>**Due**:<font color=black> Sat Mar 4 11:59 PM

<font color=red>**NOTE:**<font color=black> These questions are intended to get your feet wet in the spectrum and **breadth** of open research problems in NLP. They are *not* intended for you to explore a problem in any **depth**.

For the coding questions, aim to have a simple, working prototype that you could revisit at a later time if you wanted to improve its performance.
    
A "best practice" is to work these out *by hand* first to know what is reasonable for your code to be able to do (knowing it can be done by a human). An example is identifying whether the word `run` is a verb or not. Can you determine whther `run` is a verb?
    
Note that some of the tasks in this assignment could be considered a Master's Thesis (if not a PhD thesis)! Do not get frustrated or overwhelmed by the difficulty of each problem.
    
Feel free to work with others but make sure you are submitting your own work.

---
<font color=blue>
    
#### OBJECTIVES
    
<font color=black>

* become familiar with the `JupyterLab Notebook` format for the assignments
* embrace failure
* learn from failure, don't just give up when something doesn't work
* get your feet wet in seemingly simple text processing problems
* become familiar with `NLTK`
* encounter a variety of open problems in natural language processing *prior* to studying them formally (in order to motivate their formal study)
* practice the workflow that will be used throughout this course: gather data, implement code, evaluate code, discuss
* emphasize the **discussion** and **reflection** aspect of computing science/research

---
<font color=darkred>
    
# Preparation: NLTK Natural Language Toolkit

<font color=black>

**NLTK** stands for **Natural Language Toolkit**. It is written in **Python** and intended to be fairly easy to pick up.
    
The **NLTK book** is found here: https://www.nltk.org/book

In order to complete this assignment you must complete Chapters #1, #2, and #3. It is recommended you not just read through the examples but create a **Jupyter Notebook** where you are typing the various commands as you go.

A few of your other assignments will use **NLTK** as well.

---
<font color=darkred>
    
# Question: Tokenizing Text

<font color=black>

You will process text saved to the `text_data` variable. The variable is of type `String`.

This data is know as a (very small) **dataset**, or sometimes referred to as **corpora**.
The etymology of <font color=darkred>**corpora**<font color=black> is it comes from **corpus** or **corpse**, meaning the body of something. So **corpora** refers to a **body of texts** (or collection of texts).

In [1]:
text_data = """
Here’s to the crazy ones, the misfits, the rebels, the troublemakers, the round pegs in the square holes. The ones who see things differently - they’re not fond of rules. You can quote them, disagree with them, glorify or vilify them, but the only thing you can’t do is ignore them because they change things. They push the human race forward, and while some may see them as the crazy ones, we see genius, because the ones who are crazy enough to think
that they can change the world, are the ones who do.
The quote baove is by Steve Jobs. Mr. Jobs also said: I choose a lazy person to do a hard job. Because a lazy person will find an easy way to do it.
"""

Break the above text into **paragraph tokens** (a list of paragraphs).\
How many paragraphs do you have?

In [2]:
# YOUR CODE IN THIS CELL

paragraphs = text_data.split(".\n")
count=0
for x in paragraphs:
    if(x.strip() != '' and len(x)!= 1): 
        count+=1
#Printing the paragraphs
for x in paragraphs:
    print(x)
    print("\n")
    
print("Number of paragraph tokens:", count)


Here’s to the crazy ones, the misfits, the rebels, the troublemakers, the round pegs in the square holes. The ones who see things differently - they’re not fond of rules. You can quote them, disagree with them, glorify or vilify them, but the only thing you can’t do is ignore them because they change things. They push the human race forward, and while some may see them as the crazy ones, we see genius, because the ones who are crazy enough to think
that they can change the world, are the ones who do


The quote baove is by Steve Jobs. Mr. Jobs also said: I choose a lazy person to do a hard job. Because a lazy person will find an easy way to do it





Number of paragraph tokens: 2


Break the above text into **sentence tokens**.\
How many sentences do you have?

In [3]:
# YOUR CODE IN THIS CELL

#Using Regex Library
import re 
count_of_sentences = 0
sentences = re.split('[?.!;]', text_data)

for x in sentences:
    if(x.strip() != ''):
        count_of_sentences+=1

for x in sentences:
    print(x)
    print("\n")

print("Number of sentence tokens: ", count_of_sentences)


Here’s to the crazy ones, the misfits, the rebels, the troublemakers, the round pegs in the square holes


 The ones who see things differently - they’re not fond of rules


 You can quote them, disagree with them, glorify or vilify them, but the only thing you can’t do is ignore them because they change things


 They push the human race forward, and while some may see them as the crazy ones, we see genius, because the ones who are crazy enough to think
that they can change the world, are the ones who do



The quote baove is by Steve Jobs


 Mr


 Jobs also said: I choose a lazy person to do a hard job


 Because a lazy person will find an easy way to do it






Number of sentence tokens:  8


Break the above text into **word tokens**.\
How many words do you have?

In [4]:
# YOUR CODE IN THIS CELL
import string

translator = str.maketrans('', '', string.punctuation)
new_string = text_data.translate(translator)

new_words = new_string.split();
for x in new_words:
    x = x.strip()
for x in new_words:
    print(x)

print("Number of words tokens:", len(new_words))


Here’s
to
the
crazy
ones
the
misfits
the
rebels
the
troublemakers
the
round
pegs
in
the
square
holes
The
ones
who
see
things
differently
they’re
not
fond
of
rules
You
can
quote
them
disagree
with
them
glorify
or
vilify
them
but
the
only
thing
you
can’t
do
is
ignore
them
because
they
change
things
They
push
the
human
race
forward
and
while
some
may
see
them
as
the
crazy
ones
we
see
genius
because
the
ones
who
are
crazy
enough
to
think
that
they
can
change
the
world
are
the
ones
who
do
The
quote
baove
is
by
Steve
Jobs
Mr
Jobs
also
said
I
choose
a
lazy
person
to
do
a
hard
job
Because
a
lazy
person
will
find
an
easy
way
to
do
it
Number of words tokens: 126


---

<font color=darkred>
    
# Gathering Results: Frequency Distributions
    
<font color=black>

Determine what the most frequent words in the `text_data` variable are. List the top 20 most frequent words.

In [5]:
# YOUR CODE IN THIS CELL
import string

#Removing the punctuations
translator = str.maketrans('', '', string.punctuation)
new_string = text_data.translate(translator)

#Split the text_data into a list
new_words = new_string.split();

#Changing all the words into lower case
lowercase_list = [element.lower() for element in new_words]

#Creating a dictionary to store the word frequency
word_freq = {}

for word in lowercase_list:
    if word in word_freq:
        word_freq[word] += 1
    else:
        word_freq[word] = 1
        
sorted_word_freq = sorted(word_freq.items(), key=lambda x: x[1], reverse=True)

#Getting 20 most frequent words
most_frequent_20_words = [word[0] for word in sorted_word_freq[:20]]

print("The top 20 most frequent words are:")
for words in most_frequent_20_words:
    print(words)

The top 20 most frequent words are:
the
ones
them
to
do
crazy
who
see
because
they
a
things
you
can
quote
is
change
are
jobs
lazy


---

<font color=darkred>
    
# Reflection & Thoughts
    
<font color=black>

Discuss the following:
* how well did your code do at tokenizing the text into words?
* how did you evaluate whether it performed well?
* is there anything surprising or unusual in your findings/results?
* why did you get the results you did?
* what can you do better to improve the performance/results?
* what appears to potentially provide the biggest boost in performance if you could have taken care of the shortcoming?
* how represent is the dataset of text data in general?

Reflection and Thoughts: 

Paragraph Tokens:
In the case, we are tokenizing the text into paragraph tokens, we split the text into paragraph(list) at ".\n" using .split function and then returning the count of the elements in the list that are non-empty and have more than one character which means we are assuming each paragraph ends with a period followed by a newline character.I evaluated the code by a couple of test cases including the one provided(text_data). The code runs perfectly for the given text_data (test case)and returned 2 as the number of paragraphs in the text_data because all the paragraphs in the text_data ends with period followed by newline character which is usually the case.But it is not always the case. The paragraph can even end with exclamation point(!), question mark(?), semicolon(;) etc.To handle these cases and to improve the performance/results, we need to optimize our program to handle the case where paragraphs end with !,?, ; etc. More sophisticated text processing techniques, such as regular expressions or natural language processing libraries like NLTK will be needed to successfully handle those cases. This could potientially be the biggest boost in the performance. I think if the dataset had few more paragraphs that ended with exclamation point(!), question mark(?), semicolon(;) and other punctuation marks then it would have been a good text data in general.

Sentence Tokens: 
In this case, we are tokenizing the text into sentences, we split the text into sentences using RegEx library to identify sentence-ending punctuations marks like period, exclamation mark, semicolon and question mark.I evaluated the code by a couple of test cases including the one provided(text_data).The code did well in splitting the text into sentences.But there are still some inconsistencies like in the text_data, there are some abbreviation like Mr. which contains period that does not necessarily indicate end of sentence which resulted in our dataset to split into one extra token i.e. 8 sentence tokens instead of 7 sentence tokens. Also, some sentences may not even end with one of the punctuation marks specified in the regular expression, such as sentences that end with a dash or a closing quotation mark. To successfully handle these cases and to improve the performance/results, more advanced natural language processing techniques, such as using a pre-trained model is required. 

Word Tokens:
In this case, we are tokenizing the text into words, we first removed the punctuations from the given test_data so that punctuation marks are not counted as separate words and then split it into word tokens. The program returns 126 word tokens in total(excluding the punctuation marks) which is practically highly accurate model. I evaluate the performance of the code using several test cases. The program might not handle non-ASCII characters properly like words with accents or non-English characters may be split incorrectly or not at all. Also, the code does not remove numbers or other non-alphabetic characters, so they will be treated as separate words. Taking care of these short-coming will provide a big-boost to the performance/results.

Frequency Distributions: 
In this case, we are listing 20 most frequent words in the sample data. We first removed punctuations from sample data( text_data) to avoid punctuations being counted as words, then spliting the entire text into tokens, changing them to lower-case to avoid words like "The" and "the" being conisdered as different words and inserting them in a dictionary to keep the count of each word. I evaluated the performance by testing the code with several test cases. The code works perfectly for the given test data and returns ['the', 'ones', 'them', 'to', 'do', 'crazy', 'who', 'see', 'because', 'they', 'a', 'things', 'you', 'can', 'quote', 'is', 'change', 'are', 'jobs', 'lazy'] as the 20 most frequent words in the text_data. 
The possible shortcoming in our code is that it is not optimized for large datasets, as it creates a dictionary to store the word frequency, which could consume a significant amount of memory and it won't be a good/efficient solution for large datasets. If we are able to tackle this problem it would be potiential boost in the performance of our code.


---

<font color=#B6321C>
    
### RUBRIC

<font color=black>

I will be evaluating this assignment mostly by your attempt to explore and probe the problem (via code or via curiousity). The assignments will build on developing critical thinking and <font color=darkgreen>**asking questions**<font color=black>,  and <font color=darkred>*not coming up with the answers*<font color=black>.

A breakdown of the marking for this assignment:
* <font color=#B6321C>[**15 marks**]<font color=black> for tokenizing the text:
    * <font color=#B6321C>[**5 marks**]<font color=black> for tokenizing paragraphs
    * <font color=#B6321C>[**5 marks**]<font color=black> for tokenizing sentences
    * <font color=#B6321C>[**5 marks**]<font color=black> for tokenizing words
* <font color=#B6321C>[**10 marks**]<font color=black> for presenting the frequency distributions of the top 20 most frequent words
* <font color=#B6321C>[**25 marks**]<font color=black> for analysis of the experiment, thoughts about the results acquired, ideas for future work or what could possibly improve the current system's performance, general reflective thoughts

The total of this assignment is <font color=#B6321C>[**50 marks**]<font color=black>. It will be scaled to be out of 10 on **Blackboard**.\
Most students will get close to perfect.
    
If you need help, feel free to email me any questions and/or post your questions to <font color=#B6321C>**Discord**<font color=black>.

---
<font color=darkred>
    
# QUESTION: Creating Rules To Identify Verbs
    
<font color=black>

Here is a list of **verbs** taken from https://eslgrammar.org/list-of-verbs:

> Accept
Achieve
Add
Admire
Admit
Adopt
Advise
Agree
Allow
Announce
Appreciate
Approve
Argue
Arrive
Ask
Assist
Attack
Bake
Bathe
Be
Beat
Become
Beg
Behave
Bet
Boast
Boil
Borrow
Breathe
Bring
Build
Burn
Bury
bury
Buried
Burying
Buy
Call
Catch
Challenge
Change
Cheat
Chew
Choose
Clap
Clean
Collect
Compare
Complain
Confess
Confuse
Construct
Control
Copy
Count
Create
Cry
Damage
Dance
Deliver
Destroy
Disagree
Drag
Drive
Drop
Earn
Eat
Employ
Encourage
Enjoy
Establish
Estimate
Exercise
Expand
Explain
Fear
Feel
Fight
Find
Fly
Forget
Forgive
Fry
Gather
Get
Give
Glow
Greet
Grow
Guess
Harass
Hate
Hear
Help
Hit
Hope
Identify
Interrupt
Introduce
Irritate
Jump
Keep
Kick
Kiss
Laugh
Learn
Leave
Lend
Lie
Like
Listen
Lose
Love
Make
Marry
Measure
Meet
Move
Murder
Obey
Offend
Offer
Open
Paint
Pay
Pick
Play
Pray
Print
Pull
Punch
Punish
Purchase
Push
Quit
Race
Read
Relax
Remember
Reply
Retire
Rub
See
Select
Sell
Send
Sing
Snore
Stand
Stare
Start
Stink
Study
Sweep
Swim
Take
Talk
Teach
Tear
Tell
Thank
Travel
Type
Understand
Use
Visit
Wait
Walk
Want
Warn
Wed
Weep
Wink
Worry
Write
Yell

The above list of verbs is not intended to be comprehensive.

<div class="alert alert-danger">
    <p>Note that you are not expected to use any actual algorithms or existing code/tools to perform this task. You are expected to approach the task naively and begin from scratch.</p>
</div>

#### READING FROM A FILE

Copy the above list of verbs into a file called `list_of_verbs.txt`. Save this file at the same location as `Assignment-2.ipynb`, then execute the following code cell.

In [6]:
def read_list_of_verbs_from_file():
    verb_file = open('list_of_verbs.txt')
    file_contents = verb_file.read()
    verb_file.close()
    return file_contents.split("\n")

list_of_verbs = read_list_of_verbs_from_file()

print(list_of_verbs) 

['Accept Achieve Add Admire Admit Adopt Advise Agree Allow Announce Appreciate Approve Argue Arrive Ask Assist Attack Bake Bathe Be Beat Become Beg Behave Bet Boast Boil Borrow Breathe Bring Build Burn Bury bury Buried Burying Buy Call Catch Challenge Change Cheat Chew Choose Clap Clean Collect Compare Complain Confess Confuse Construct Control Copy Count Create Cry Damage Dance Deliver Destroy Disagree Drag Drive Drop Earn Eat Employ Encourage Enjoy Establish Estimate Exercise Expand Explain Fear Feel Fight Find Fly Forget Forgive Fry Gather Get Give Glow Greet Grow Guess Harass Hate Hear Help Hit Hope Identify Interrupt Introduce Irritate Jump Keep Kick Kiss Laugh Learn Leave Lend Lie Like Listen Lose Love Make Marry Measure Meet Move Murder Obey Offend Offer Open Paint Pay Pick Play Pray Print Pull Punch Punish Purchase Push Quit Race Read Relax Remember Reply Retire Rub See Select Sell Send Sing Snore Stand Stare Start Stink Study Sweep Swim Take Talk Teach Tear Tell Thank Travel T

#### CREATING A SET OF RULES

Use **Table 4.2** (from https://www.nltk.org/book/ch01.html) to create some potential rules/characteriztics for identifying whether a word is a **verb**. For example, if the word ends in `ing` then it can be considered a verb. Another simple rule to implement is to see if the word is in our `list_of_verbs`.

In [9]:
import nltk
from nltk.corpus import wordnet

def is_word_a_verb(word):
        with open('list_of_verbs.txt') as file:
            list_of_verbs = file.read().lower()
            list_of_verbs = list_of_verbs.split()
        if word.lower() in list_of_verbs:
            return True
        
        if word[:-3] in nltk.corpus.words.words() and word.endswith('ing'):
            return True
        
        if  word[:-2] in nltk.corpus.words.words() and word.endswith('es'):
            return True
        
        if word[:-1] in nltk.corpus.words.words() and word.endswith('s'):
            return True
        
        given_word = nltk.pos_tag([word])[0]
        if given_word[1].startswith('VB'):
            return True
        return False
    
for x in testing_data:
    if(is_word_a_verb(x)== True):
        print(f"{x} is a verb")
    else:
        print(f"{x} is not a verb")

hat is not a verb
run is a verb
sadness is not a verb
crying is a verb
chew is a verb
kiss is a verb


#### DATA TO TEST PERFORMANCE OF CODE

An example of what the test data could look like is the following list of words:

In [8]:
testing_data = [
    "hat",
    "run",
    "sadness",
    "crying",
    "chew",
    "kiss",
]

Your code will take one word at a time from the list and determine whether the word was a verb or not. For example:

`is_word_a_verb("hat")`

would return `False`, whereas:

`is_word_a_verb("crying")`

would return `True`.

Note that the list of words in `testing_data` are unrelated to each other, meaning they are not necessarily from the same sentence (or even the same document).

#### EVALUATION

Evaluate how well your code performs on the list of words in `testing_data`.

Add your own set of words to `testing_data` to further evaluate how robust your code is.

#### DISCUSSION

Briefly discuss the performance of your code (1-2 paragraphs).

What types of verbs did the code perform well on? What types of verbs did it not perform well on?

Are there any problems with this approach in identifying whether a word is a verb?

Are there any issues with the list of verbs in our `list_of_verbs.txt` file?

<div class="alert alert-info">
    <h4>PRO TIP</h4>
    <p>The best approach to this question is to work on <strong>one aspect at a time</strong>. Since one aspect is not going to identify all possible verbs, treat each aspect as a <em>step</em> toward a destination where improvements are likely small increments.</p>
</div>



<font color=#B6321C>
    
### RUBRIC

<font color=black>

We will be evaluating this section in part by your **ambition** in trying to capture difficult aspects of a verb, **not how well** you encoded the aspects of the verbs you were trying to capture.\
A breakdown of the marking for this task:
* <font color=#B6321C>[**5 marks**]<font color=black> basic features gathered (trivial, used one or two builtin `NLTK`/`Python` methods in a simple way)
* <font color=#B6321C>[**5 marks**]<font color=black> distinctly different aspects gathered (advanced), corresponding to aspects not exactly trivial to encode (tried using a few builtin `NLTK`/`Python` methods in a complex way)
* <font color=#B6321C>[**5 marks**]<font color=black> discussion of results explaining the performance of the code

In [ ]:
Rules to Identify Verbs:
    The program for identifying the verbs performs well on the verbs that are in the list_of_verbs. It correctly identifies 
    and returns True if the word is present in the list_of_verbs. It did not work perfectly on few cases
    where the given word is not in the list of verbs or in the NLTK word corpus, in such cases, it returns False, which may 
    not always be accurate.
    The potiential problem with this approach is that It uses NLTK to check if the word is a verb, 
    which may not always be accurate or appropriate for all contexts.It also does not account for irregular verbs or the verbs 
    with unusual spellings. This issue with list_of_verbs.txt is that this file does not all the verbs that exist (which is
    practically to possible) so there might be the case that the word we take as argument in the function is_word_a_verb is 
    not present in the file. Also, accept will be different from Accept because the program would considered both
    of them to be different. To tackle this problem we need to first convert the word to lower-case and then conpare. 
    The code is efficient and detects most of the verb correctly.
    
                                                                                                    


---

<font color=darkred>
    
# QUESTION: Identifying Proper Names In A Text Document
    
<font color=black>
    

In [ ]:
text_data = """
Here’s to the crazy ones, the misfits, the rebels, the troublemakers, the round pegs in the square holes. The ones who see things differently — they’re not fond of rules. You can quote them, disagree with them, glorify or vilify them, but the only thing you can’t do is ignore them because they change things. They push the human race forward, and while some may see them as the crazy ones, we see genius, because the ones who are crazy enough to think
that they can change the world, are the ones who do.
The quote baove is by Steve Jobs. Mr. Jobs also said: I choose a lazy person to do a hard job. Because a lazy person will find an easy way to do it. Steve started the apple computer company with Steve Wozniak (aka `Woz'). Little known facts
Trump did not like Jobs. Jobs did not like Trump.
"""

Similar to the previous question ("*Creating Rules To Identify Verbs*"), write code that automatically identifies all words/terms that are **proper names**. Use the sample text provided in `text_data` above.

<div class="alert alert-danger">
    <p>Note that you are not expected to use any actual algorithms or existing code/tools to perform this task. You are expected to approach the task naively and begin from scßratch.</p>
</div>

In [ ]:
import nltk

words = nltk.word_tokenize(text_data)
tag = nltk.pos_tag(words)
entities = nltk.ne_chunk(tag, binary=True)

for entity in entities:
    if hasattr(entity, 'label') and entity.label() == "NE":
        print(" ".join(word for word, tag in entity.leaves()))



Examples of **proper names** are:
* John
* Vancouver
* Trudeau
* Pierre Trudeau
* British Columbia, Canada
* Rememberance Day

In [ ]:

Performance of the Code:
    The program identifies proper names in the text document. It is designed to perform entity recogition(NER) on the text and
    extract proper names. The program performs well on recognizing and extracting proper names that are labeled as NE such as 
    people, organization, and locations.The program does not perform well on other types of proper names, such as brand names,
    product names and scientific terms/names.
    The code is designed to perform named entity recognition (NER) on the text and extract proper names.It performs well on 
    recognizing and extracting proper names that are labeled as named entities (NE) in the text, such as names of people, 
    organizations, and locations.The code may not perform well on recognizing and extracting other types of proper names, 
    such as brand names, product names, and scientific terms. These types of proper names are not labeled as NE in the text and
    may require additional techniques to identify and extract them accurately. 
    Additionally, the code may also make errors in identifying NEs, such as recognizing common nouns or adjectives as NEs or 
    missing some NEs that are mentioned in the text.

#### DISCUSSION

Briefly discuss the performance of your code (1-2 paragraphs).

What types of proper names did the code perform well on? What types of proper names did it not perform well on?

<font color=#B6321C>
    
### RUBRIC

<font color=black>

Again, we will be evaluating this section for your **ambition** in trying to capture different aspects of a proper name, **not how well** you coded the aspects of the proper names.\
A breakdown of the marking for this task:
* <font color=#B6321C>[**5 marks**]<font color=black> basic features gathered (trivial, used one or two builtin `NLTK`/`Python` methods in a simple way)
* <font color=#B6321C>[**5 marks**]<font color=black> distinctly different aspects gathered (advanced), corresponding to aspects not exactly trivial to encode (tried using a few builtin `NLTK`/`Python` methods in a complex way)
* <font color=#B6321C>[**5 marks**]<font color=black> discussion of results explaining the performance of the code